pkg/apis/resource/types.go

```diff
@@ -82,6 +82,23 @@ const (
 
 // ResourceSliceSpec contains the information published by the driver in one ResourceSlice.
 type ResourceSliceSpec struct {
+	// RequiredVersion defines which semantic of a ResourceSliceSpec must
+	// be supported by clients.
+	//
+	// Clients must check this field and ignore all ResourceSlices with a
+	// version that is higher than the one they support. Drivers should use
+	// the lowest version that provides the features they need.
+	//
+	// The version follows semver 2.0.0.
+	//
+	// History:
+	// - 1.31.0: initial definition of simple devices with their own attributes.
+	// - 1.32.0: partitionable devices, default attributes inherited by
+	//   devices from the slice
+	//
+	// +required
+	RequiredVersion string
```

Comment:
If we want different semantic without the `RequiredVersion`, then in the Kubernetes model we would need different top-level kinds ("ResourceSlice", "PartitionableResourceSlice", etc.) but then all of them need to be supported concurrently and clients need to watch all, increasing overhead.

Reply:
Is there precedent for this elsewhere? It seems like a new concept.

Here's another thought (sorry for one more option!). Rather than thinking of this as "types" or "versions", perhaps another way to think of it is as API "features" or "capabilities":

- Basic devices with attributes
- Partitionable devices
- Devices with consumable resources
- something in the future, like "access modes", etc.

In this way, rather than a single string discriminator, we would instead have an array of features used. Since resource slices are published by drivers and not human-crafted, this probably wouldn't be too annoying. For example:

```go

// ResourceSliceFeature identifies control plane features needed for this resource slice
// to be properly managed. Valid values:
// - devices
// - partitionable
// - consumable-resources
type ResourceSliceFeature string

type ResourceSliceSpec struct {

// FeaturesUsed lists the features that a Kubernetes control plane must implement in
// order to properly manage the resources in this slice. If an unknown feature is seen, then
// the control plane should ignore this slice.
FeaturesUsed []ResourceSliceFeature

...

```

Reply:
Basically I think we have a multi-dimensional problem and "types" or "versions" is a single-dimensional concept...

Yes this is also not something we do elsewhere, though we do have similar concepts in other places (NFD, CRI features, I suspect Gateway probably does something like this somewhere...)

Reply:
> Is there precedent for this elsewhere? It seems like a new concept.

It's probably new.

> another way to think of it is as API "features" or "capabilities":

A slice of feature flags also works. The downside is that it's harder to reason about compatibility: instead of a simple version comparison ("does component XYZ support ResourceSlices with required version 1.33"?) one has to check support for individual features.

I don't have a strong opinion either way.

Reply:
Tim and I were just chatting and a couple things came up:
- If partitionable devices is a sibling to device, as in the [other PR](https://github.com/kubernetes/kubernetes/pull/125952), then it is more analogous to a "gizmo" - that probably implies completely different requests and classes, etc. It doesn't have to, I suppose. A `DeviceRequest` could be satisfied by a "device" or "partitionable device" but not a "gizmo". But that does mean there is a some implicit hard coded mapping between type of request (device, gizmo), and different resource slice one-ofs (device, partitionabledevice, gizmo), rather than just a simple 1:1 relationship. That puts another "+1" for me on this PR's approach.
- Similarly, features like the ones listed above apply across devices and partitionable devices, but not necessarily across gizmos. Splitting the devices and partitionable devices into separate fields *might* make that harder to implement.
- These PRs consider how the evolution of the `ResourceSlice` API works, but we also need to think about how the evolution of the `DeviceRequest` ties to those. For example, if we add `requests` to ask for per-device allocatable resources, an older scheduler will not see that field. Instead, it will allocate a whole device, unless it somehow knows not to process that request. If the device has enough of the resource, that is technically OK. But it won't be checking that since it ignores `requests` altogether.
- There may also be a relationship between the claim features and the slice features which we need to understand more clearly. For example, a scheduler could check the slice features and ignore slices that have features that it doesn't understand. But that might not be what we want either - if you upgrade the driver and it supports some new feature, the scheduler may stop processing those slices! 

I'll start thinking through these and put an issue in https://github.com/kubernetes-sigs/wg-device-management/issues.

Reply:
> it is more analogous to a "gizmo" - that probably implies completely different requests and classes

I don't agree here. For a user, the way to ask for a partition of a GPU or a full GPU is exactly the same. They might not even care what they get, as long as it has certain attributes.

> we also need to think about how the evolution of the DeviceRequest ties to those. For example, if we add requests to ask for per-device allocatable resources, an older scheduler will not see that field.

For CEL, we have pretty good versioning support. We can add new constructs that will cause compile errors in older clients. Constraints are already a one-of. `DeviceRequest` isn't yet, but could become one - that is the discussion in https://github.com/kubernetes/kubernetes/pull/125488#discussion_r1667773527. So I think we are in an okay(ish) state on the request side.

> if you upgrade the driver and it supports some new feature, the scheduler may stop processing those slices!

I don't think there is a way around that. If a driver depends on new features, then the scheduler has to support them. A driver can mitigate this a bit by publishing one ResourceSlice with "simple devices" and another with the more complex devices. Then only the complex devices get ignored. But that might not be possible, depending the devices.

Reply:
I don't think @johnbelamaric is saying it **should** be treated more like a gizmo, but by creating a new field at the top-level as an alternate to `Devices` it makes it it looks like it could/should be created as a gizmo (since that is also how a new gizmo would be introduced).

Reply:
Perhaps. OTOH, if it's called `<Something>Device` then to me there's no such confusion.

@johnbelamaric : regarding adding a field to a `DeviceRequest` - I reopened https://github.com/kubernetes/kubernetes/pull/125488#discussion_r1669241707 and added that comment there.

Reply:
Another variant.

Device.Schema is only needed because there are no required fields in a leaf
device -- it could have no attributes and no capacities, right?

```
type ResourceSliceSpec struct {
	// other stuff

	Devices []Device
}

type DeviceSchema string

const (
	DeviceSchemaSingle DeviceSchema = "SingleDevice"
	DeviceSchemaMacro DeviceSchema = "MacroDevice"
)

type Device struct {
	// Name is unique identifier among all devices managed by
	// the driver in the pool. It must be a DNS label.
	//
	// +required
	Name string

	// Schema indicates how this device is modelled in this API.
	//
	// +optional
	// +default=Leaf
	Schema DeviceSchema

	// Attributes defines the set of attributes for this device.
	// The name of each attribute must be unique in that set.
	// If schema is "MacroDevice", then these attributes apply to all
	// sub-devices.
	//
	// The maximum number of attributes and capacities combined is 32.
	//
	// +optional
	Attributes map[QualifiedName]DeviceAttribute

	// The following fields only apply when schema is "LeafDevice".

	// Capacity defines the set of capacities for this device.
	// The name of each capacity must be unique in that set.
	//
	// The maximum number of attributes and capacities combined is 32.
	//
	// +optional
	Capacity map[QualifiedName]resource.Quantity

	// The following fields only apply when schema is "MacroDevice".

	// SharedCapacity defines the set of shared capacity consumable by
	// leaf devices.
	//
	// Must not have more than 128 entries. Names must be C identifiers
	// that are unique inside the ResourceSlice.
	//
	// +optional
	SharedCapacity map[string]resource.Quantity

	// Devices lists some or all of the partitionable devices in this pool.
	// Partitionable devices may inherit shared attributes from the
	// ResourceSlice and they may (but don't have to) consume shared
	// capacity.
	//
	// +optional
	// +listType=atomic
	Devices []PartitionedDevice
}

type PartitionedDevice struct {
	// Attributes defines the set of attributes for this device.
	// The name of each attribute must be unique in that set.
	//
	// The maximum number of attributes and capacities combined is 32.
	//
	// +optional
	Attributes map[QualifiedName]DeviceAttribute

	// Capacity defines the set of capacities for this device.
	// The name of each capacity must be unique in that set.
	//
	// The maximum number of attributes and capacities combined is 32.
	//
	// +optional
	Capacity map[QualifiedName]resource.Quantity

	// SharedCapacityConsumed defines the set of shared capacity consumed by
	// this device.
	//
	// Must not have more than 32 entries.
	//
	// +optional
	SharedCapacityConsumed map[string]resource.Quantity
}
```

We could make Capacity mean "shared capacity" if schema is MacroDevice?  It's
more complex, and doesn't handle the int-range stuff, so we need more fields.
I don't love it.

We could make Name be specific to SingleDevice, then we don't need schema?

```
type ResourceSliceSpec struct {
	// other stuff

	Devices []Device
}

type Device struct {
	// These fields are what we ship in 1.31 and how we model a "simple"
	// single device.

	// Name is unique identifier among all devices managed by
	// the driver in the pool. It must be a DNS label.
	//
	// +optional
	// +oneOf=Schema
	Name string

	// Attributes defines the set of attributes for this device.
	// The name of each attribute must be unique in that set.
	// This field only applies when the name field is specified.
	//
	// The maximum number of attributes and capacities combined is 32.
	//
	// +optional
	Attributes map[QualifiedName]DeviceAttribute

	// Capacity defines the set of capacities for this device.
	// The name of each capacity must be unique in that set.
	// This field only applies when the name field is specified.
	//
	// The maximum number of attributes and capacities combined is 32.
	//
	// +optional
	Capacity map[QualifiedName]resource.Quantity

	// Devices lists some or all of the partitionable devices in this pool.
	// Partitionable devices may inherit shared attributes from the
	// ResourceSlice and they may (but don't have to) consume shared
	// capacity.
	//
	// +optional
	// +listType=atomic
	// +oneOf=Schema
	Devices []PartitionedDevice

	// SharedCapacity defines the set of shared capacity consumable by
	// leaf devices.
	// This field only applies when the devices field is specified.
	//
	// Must not have more than 128 entries. Names must be C identifiers
	// that are unique inside the ResourceSlice.
	//
	// +optional
	SharedCapacity map[string]resource.Quantity
}

type PartitionedDevice struct {
	// Name is unique identifier among all devices managed by
	// the driver in the pool. It must be a DNS label.
	//
	// +required
	Name string

	// Attributes defines the set of attributes for this device.
	// The name of each attribute must be unique in that set.
	//
	// The maximum number of attributes and capacities combined is 32.
	//
	// +optional
	Attributes map[QualifiedName]DeviceAttribute

	// Capacity defines the set of capacities for this device.
	// The name of each capacity must be unique in that set.
	//
	// The maximum number of attributes and capacities combined is 32.
	//
	// +optional
	Capacity map[QualifiedName]resource.Quantity

	// SharedCapacityConsumed defines the set of shared capacity consumed by
	// this device.
	//
	// Must not have more than 32 entries.
	//
	// +optional
	SharedCapacityConsumed map[string]resource.Quantity
}
```

Or we go back to the level of indirect we tried to elide before.  We have an
opportunity here in that we have a decent comprehension of the near future
(better than we usually do, at least), so we can make some decisions NOW that
make LATER easier.  This is epsilon uglier today, but way less abiguous

```
type ResourceSliceSpec struct {
	// other stuff

	Devices []Device
}

type Device struct {
	// optional
	// +oneOf=Device
	Single *SingleDevice

	// optional
	// +oneOf=Device
	Macro []MacroDevice
}

type SingleDevice struct {
	// Name is unique identifier among all devices managed by
	// the driver in the pool. It must be a DNS label.
	//
	// +required
	Name string

	// Attributes defines the set of attributes for this device.
	// The name of each attribute must be unique in that set.
	// This field only applies when the name field is specified.
	//
	// The maximum number of attributes and capacities combined is 32.
	//
	// +optional
	Attributes map[QualifiedName]DeviceAttribute

	// Capacity defines the set of capacities for this device.
	// The name of each capacity must be unique in that set.
	// This field only applies when the name field is specified.
	//
	// The maximum number of attributes and capacities combined is 32.
	//
	// +optional
	Capacity map[QualifiedName]resource.Quantity
}

type MacroDevice struct {
	// Devices lists some or all of the partitionable devices in this pool.
	// Partitionable devices may inherit shared attributes from the
	// ResourceSlice and they may (but don't have to) consume shared
	// capacity.
	//
	// +optional
	// +listType=atomic
	// +oneOf=Schema
	Devices []LeafDevice

	// SharedCapacity defines the set of shared capacity consumable by
	// leaf devices.
	// This field only applies when the devices field is specified.
	//
	// Must not have more than 128 entries. Names must be C identifiers
	// that are unique inside the ResourceSlice.
	//
	// +optional
	SharedCapacity map[string]resource.Quantity
}

type LeafDevice struct {
	// Name is unique identifier among all devices managed by
	// the driver in the pool. It must be a DNS label.
	//
	// +required
	Name string

	// Attributes defines the set of attributes for this device.
	// The name of each attribute must be unique in that set.
	//
	// The maximum number of attributes and capacities combined is 32.
	//
	// +optional
	Attributes map[QualifiedName]DeviceAttribute

	// Capacity defines the set of capacities for this device.
	// The name of each capacity must be unique in that set.
	//
	// The maximum number of attributes and capacities combined is 32.
	//
	// +optional
	Capacity map[QualifiedName]resource.Quantity

	// SharedCapacityConsumed defines the set of shared capacity consumed by
	// this device.
	//
	// Must not have more than 32 entries.
	//
	// +optional
	SharedCapacityConsumed map[string]resource.Quantity
}
```

Reply:
So the last option is basically `Device struct` as a one-of, a concept that we understand well.

The rule for a consumer would be "ignore all unknown device types" instead of the "ignore the whole slice" that we would have to use with the discriminator or feature set approach.

I think this works. Because ResourceSlice is primarily used between machines, the additional indirection is not much of a burden.

:+1: from me.

Reply:
I would call the struct in 1.31 `SimpleDevice` instead of `SingleDevice`. It seems safe to assume that whatever comes next is more complex.

Reply:
If I don't hear otherwise, then I'll implement this approach in https://github.com/kubernetes/kubernetes/pull/125488 tomorrow.

Reply:
I'm looking to @johnbelamaric and @klueska for thoughts - the discussion around YAML output was a good discussion, and I want to do our best to align to that.

Reply:
This is compatible with what we are planning, it gives us enough flexibility to do that. I think we can consider adding slice level attributes in 1.31, too. That is also not incompatible with what we are thinking. But it can come in a follow-on PR, I think.

Reply:
"This" meaning the one-of. The specific implementation in this RFC PR is not the one we will use for partitionable.

Reply:
Yes. Doing things this way is flexible enough for us to extend things for partitionable devices (and beyond) without sacrificing the ability to represent "simple" devices as we have defined them today.